accounting/calculated.go

```diff
@@ -0,0 +1,151 @@
+}
+
+func convertAssetHoldings(gAssets *[]generated.AssetHolding) map[basics.AssetIndex]basics.AssetHolding {
+	// assumes WYSIWYG ignoring round information except for Deleted which is skipped
+	// TODO: handle OptedInAtRound and OptedOutAtRound
+	// Make sure to add an integration test case that opts in and out and also back in
+	if gAssets == nil {
+		return nil
+	}
+	bAssets := make(map[basics.AssetIndex]basics.AssetHolding)
+	for _, gAsset := range *gAssets {
+		if boolZerovalIfNil(gAsset.Deleted) {
+			continue
+		}
+
+		bAssetId := basics.AssetIndex(gAsset.AssetId)
+		if _, alreadyExists := bAssets[bAssetId]; alreadyExists {
+			// skip dupes
+			continue
+		}
+
+		// for the purposes if MinBalance calculation, there is no need for Amount field, etc.
+		bAssets[bAssetId] = basics.AssetHolding{}
+	}
+	if len(bAssets) > 0 {
+		return bAssets
+	}
+	return nil
+}
+
+func convertAppsCreated(gApps *[]generated.Application) map[basics.AppIndex]basics.AppParams {
+	// assumes WYSIWYG ignoring most round information except for Deleted or DeletedAtRound which are skipped
+	if gApps == nil {
+		return nil
+	}
+	bAapps := make(map[basics.AppIndex]basics.AppParams)
+	for _, gApp := range *gApps {
+		if boolZerovalIfNil(gApp.Deleted) || gApp.DeletedAtRound != nil {
+			continue
+		}
+		bAppId := basics.AppIndex(gApp.Id)
+		if _, alreadyExists := bAapps[bAppId]; alreadyExists {
+			// skip dupes
+			continue
+		}
+
+		// MinBalance doesn't dig into AppParams, so provide a zeroval struct
+		bAapps[bAppId] = basics.AppParams{}
+	}
+	if len(bAapps) > 0 {
+		return bAapps
+	}
+	return nil
+}
+
+func convertAppsOptedIn(gApps *[]generated.ApplicationLocalState) map[basics.AppIndex]basics.AppLocalState {
+	// assumes WYSIWYG ignoring most round information except for Deleted or ClosedOutAtRound which are skipped
+	if gApps == nil {
+		return nil
+	}
+	bAapps := make(map[basics.AppIndex]basics.AppLocalState)
+	for _, gApp := range *gApps {
+		if boolZerovalIfNil(gApp.Deleted) || gApp.ClosedOutAtRound != nil {
+			continue
+		}
+		bAppId := basics.AppIndex(gApp.Id)
+		if _, alreadyExists := bAapps[bAppId]; alreadyExists {
+			// skip dupes
+			continue
+		}
+
+		// MinBalance doesn't dig into AppLocalStates, so provide a zeroval struct
+		bAapps[bAppId] = basics.AppLocalState{}
+	}
+	if len(bAapps) > 0 {
+		return bAapps
+	}
+	return nil
+}
+// minBalanceProjection projects a part of a generated.Account to a basics.AccountData struct.
+// It does so for the purpose of calculating an account's minumum balance using the official
+// calculator go-algorand/data/basics/userBalance.go::MinBalance()
+// Some data which is relevant for MinBalance but duplicated is ignored. In particular:
+// * TotalAppSchema - gotten from account.AppsTotalSchema and not re-calc'ed from CreatedApps
+// * TotalExtraPages - gotten from account.AppsTotalExtraPages and not re-calc'ed from CreatedApps
+func minBalanceProjection(account *generated.Account) basics.AccountData {
```

Comment:
There is a similar function `AccountToAccountData` in `go-algorand` that fills out all of the fields (but has some minor issues when applied to Indexer's `generated.Account`. It is also possible to try and unmarshall the `basics.Account` directly from the database as is partially done elsewhere. After some thought, I decided that this was the simplest approach - get only the essential data into a `basics.AccountData` at the most convenient location and calculate `MinBalance` from it.